gcp/api/server.py

```diff
@@ -850,8 +850,13 @@ def query_by_version(context: QueryContext,
     bugs.extend(new_bugs)
 
   # Try querying by generic version for all cases.
-  new_bugs, _ = yield _query_by_generic_version(context, query, package_name,
-                                                ecosystem, purl, version)
+
+  new_bugs, generic_page_token = yield _query_by_generic_version(
+      context, query, package_name, ecosystem, purl, version)
+
+  if ecosystem and not is_semver:
```

Comment:
We could potentialy do something like this to try and still support page tokens in the majority of the SemVer cases:

```python
if bugs and new_bugs:
   # Both queries returned results, we can't support a page token. 
  next_page_token = None
else if new_bugs:
  # The first (SemVer) query returned no results, so we can return a page token here.
  next_page_token = generic_page_token
# Otherwise, the first SemVer query returned results and the generic query returned no results. The page token should already be set to the one from the SemVer query. 
```

This also changes the behaviour slightly for the no-ecosystem-specified case in that we may return a page token in those cases, instead of never returning a page token. 

However, this will only work if either:
- We can distinguish which query a page token is for, or
- Passing an incompatible page token to a query is the same as passing no token (and there are no errors etc raised)

If none of these are possible, we may have to remove page tokens completely for SemVer queries. 

@another-rex @michaelkedar  WDYT?

Reply:
IIRC 2 is not the case, it will throw an exception if called with the wrong page token.

Reply:
Hmm, in that case, we either: 
1. ignore wrong page token exceptions and proceed with the next query. If both queries raise exceptions, then we re-raise the error. 
2. we remove paging for SemVer completely. we can just move the _query_by_generic_version call inside the query_by_semver function, and remove the page token completely. 

Let's give 1 a shot and see if it works? If not, or if it's too brittle, we can just go with 2.

Reply:
I think we can check if an ecosystem is set or not to decide if we return a `None` page token?